### PR TITLE
schedule: avoid data race and panic about scheduler controller

### LINKED ISSUE #7106: mcs: data race about scheduler controller
### What did you do?

run test

### What did you see instead?

```
2023-09-18T08:11:45.8089852Z ==================
2023-09-18T08:11:45.8089957Z WARNING: DATA RACE
2023-09-18T08:11:45.8090106Z Write at 0x00c002fc2b10 by goroutine 65006:
2023-09-18T08:11:45.8090234Z   runtime.racewrite()
2023-09-18T08:11:45.8090397Z       <autogenerated>:1 +0x1e
2023-09-18T08:11:45.8090616Z   github.com/tikv/pd/pkg/schedule/schedulers.(*Controller).Wait()
2023-09-18T08:11:45.8090998Z       /home/runner/work/pd/pd/pkg/schedule/schedulers/scheduler_controller.go:70 +0x1a4
2023-09-18T08:11:45.8091229Z   github.com/tikv/pd/pkg/schedule.(*Coordinator).RunUntilStop()
2023-09-18T08:11:45.8091552Z       /home/runner/work/pd/pd/pkg/schedule/coordinator.go:374 +0x13e
2023-09-18T08:11:45.8091812Z   github.com/tikv/pd/pkg/mcs/scheduling/server.(*Server).startCluster.func2()
2023-09-18T08:11:45.8092142Z       /home/runner/work/pd/pd/pkg/mcs/scheduling/server/server.go:447 +0x33
2023-09-18T08:11:45.8092154Z 
2023-09-18T08:11:45.8092323Z Previous read at 0x00c002fc2b10 by goroutine 65004:
2023-09-18T08:11:45.8092444Z   runtime.raceread()
2023-09-18T08:11:45.8092608Z       <autogenerated>:1 +0x1e
2023-09-18T08:11:45.8092854Z   github.com/tikv/pd/pkg/schedule/schedulers.(*Controller).AddScheduler()
2023-09-18T08:11:45.8093230Z       /home/runner/work/pd/pd/pkg/schedule/schedulers/scheduler_controller.go:187 +0x2dd
2023-09-18T08:11:45.8093479Z   github.com/tikv/pd/pkg/mcs/scheduling/server.(*Cluster).updateScheduler()
2023-09-18T08:11:45.8093940Z       /home/runner/work/pd/pd/pkg/mcs/scheduling/server/cluster.go:237 +0xec4
2023-09-18T08:11:45.8094224Z   github.com/tikv/pd/pkg/mcs/scheduling/server.(*Cluster).StartBackgroundJobs.func1()
2023-09-18T08:11:45.8094557Z       /home/runner/work/pd/pd/pkg/mcs/scheduling/server/cluster.go:382 +0x33
2023-09-18T08:11:45.8094567Z 
2023-09-18T08:11:45.8094794Z Goroutine 65006 (running) created at:
2023-09-18T08:11:45.8095032Z   github.com/tikv/pd/pkg/mcs/scheduling/server.(*Server).startCluster()
2023-09-18T08:11:45.8095374Z       /home/runner/work/pd/pd/pkg/mcs/scheduling/server/server.go:447 +0xeec
2023-09-18T08:11:45.8095809Z   github.com/tikv/pd/pkg/mcs/scheduling/server.(*Server).startCluster-fm()
2023-09-18T08:11:45.8095978Z       <autogenerated>:1 +0x47
2023-09-18T08:11:45.8096223Z   github.com/tikv/pd/pkg/mcs/scheduling/server.(*Server).campaignLeader()
2023-09-18T08:11:45.8096556Z       /home/runner/work/pd/pd/pkg/mcs/scheduling/server/server.go:254 +0xc1e
2023-09-18T08:11:45.8096821Z   github.com/tikv/pd/pkg/mcs/scheduling/server.(*Server).primaryElectionLoop()
2023-09-18T08:11:45.8097162Z       /home/runner/work/pd/pd/pkg/mcs/scheduling/server/server.go:222 +0x3a4
2023-09-18T08:11:45.8097429Z   github.com/tikv/pd/pkg/mcs/scheduling/server.(*Server).startServerLoop.func1()
2023-09-18T08:11:45.8097760Z       /home/runner/work/pd/pd/pkg/mcs/scheduling/server/server.go:143 +0x33
2023-09-18T08:11:45.8097769Z 
2023-09-18T08:11:45.8097905Z Goroutine 65004 (running) created at:
2023-09-18T08:11:45.8098166Z   github.com/tikv/pd/pkg/mcs/scheduling/server.(*Cluster).StartBackgroundJobs()
2023-09-18T08:11:45.8098510Z       /home/runner/work/pd/pd/pkg/mcs/scheduling/server/cluster.go:382 +0xfc
2023-09-18T08:11:45.8098749Z   github.com/tikv/pd/pkg/mcs/scheduling/server.(*Server).startCluster()
2023-09-18T08:11:45.8099079Z       /home/runner/work/pd/pd/pkg/mcs/scheduling/server/server.go:446 +0xd3c
2023-09-18T08:11:45.8099407Z   github.com/tikv/pd/pkg/mcs/scheduling/server.(*Server).startCluster-fm()
2023-09-18T08:11:45.8099574Z       <autogenerated>:1 +0x47
2023-09-18T08:11:45.8099815Z   github.com/tikv/pd/pkg/mcs/scheduling/server.(*Server).campaignLeader()
2023-09-18T08:11:45.8100146Z       /home/runner/work/pd/pd/pkg/mcs/scheduling/server/server.go:254 +0xc1e
2023-09-18T08:11:45.8100398Z   github.com/tikv/pd/pkg/mcs/scheduling/server.(*Server).primaryElectionLoop()
2023-09-18T08:11:45.8100729Z       /home/runner/work/pd/pd/pkg/mcs/scheduling/server/server.go:222 +0x3a4
2023-09-18T08:11:45.8100998Z   github.com/tikv/pd/pkg/mcs/scheduling/server.(*Server).startServerLoop.func1()
2023-09-18T08:11:45.8101327Z       /home/runner/work/pd/pd/pkg/mcs/scheduling/server/server.go:143 +0x33
2023-09-18T08:11:45.8101417Z ==================
```

### PR DESCRIPTION
**Pull request by lhy1024**
### What problem does this PR solve?

Issue Number: Close #7106

### What is changed and how does it work?

```commit-message
```

add a flexible wait group, because WaitGroup must ensure previous Wait has returned before added again.

### Check List

Tests

- Unit test

### Release note

```release-note
None.
```

### CODE REVIEW

File: pkg/schedule/schedulers/scheduler_controller.go [hunk 1]
```diff
@@ -41,6 +41,7 @@ var denySchedulersByLabelerCounter = labeler.LabelerEventCounter.WithLabelValues
 type Controller struct {
 	sync.RWMutex
 	wg      sync.WaitGroup
+	wgLock  sync.Mutex
```
Comment (rleungx) [inline pkg/schedule/schedulers/scheduler_controller.go:44]: I think the problem is not whether we use mutex to protect it.

Comment (lhy1024) [reply pkg/schedule/schedulers/scheduler_controller.go:44]: Do you have any other ideas?

File: pkg/utils/syncutil/flexible_wait_group_test.go [hunk 2]
```diff
@@ -0,0 +1,132 @@
+// Copyright 2022 TiKV Project Authors.
```
Comment (HuSharp) [inline pkg/utils/syncutil/flexible_wait_group_test.go:1]:
```suggestion
// Copyright 2023 TiKV Project Authors.
```

File: pkg/utils/syncutil/flexible_wait_group.go [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+import (
+	"sync"
+)
+
+// FlexibleWaitGroup is a flexible version of sync.WaitGroup.
+// It supports adding to the counter after Wait() has been called,
+// which is not allowed in sync.WaitGroup.
+type FlexibleWaitGroup struct {
+	sync.Mutex
+	count int
```
Comment (HuSharp) [inline pkg/utils/syncutil/flexible_wait_group.go:26]: Do you think we can provide an interface to acquire count when exiting schedulers 🤔
Maybe it's a redundant thinking :)

**Comment by HuSharp** [comment]
seems like DATA RACE in ci
https://github.com/tikv/pd/actions/runs/6252103050/job/16974605966?pr=7107
```
==================
WARNING: DATA RACE
Write at 0x00c0000b0098 by goroutine 37:
  github.com/tikv/pd/pkg/utils/syncutil.(*FlexibleWaitGroup).Add()
      /home/runner/work/pd/pd/pkg/utils/syncutil/flexible_wait_group.go:43 +0x107
  github.com/tikv/pd/pkg/utils/syncutil.TestNegativeDelta.func2()
      /home/runner/work/pd/pd/pkg/utils/syncutil/flexible_wait_group_test.go:82 +0x[35](https://github.com/tikv/pd/actions/runs/6252103050/job/16974605966?pr=7107#step:5:36)

Previous read at 0x00c0000b0098 by goroutine 35:
  github.com/tikv/pd/pkg/utils/syncutil.TestNegativeDelta()
      /home/runner/work/pd/pd/pkg/utils/syncutil/flexible_wait_group_test.go:86 +0x224
  testing.tRunner()
      /opt/hostedtoolcache/go/1.21.1/x64/src/testing/testing.go:1595 +0x238
  testing.(*T).Run.func1()
      /opt/hostedtoolcache/go/1.21.1/x64/src/testing/testing.go:1648 +0x44

Goroutine [37](https://github.com/tikv/pd/actions/runs/6252103050/job/16974605966?pr=7107#step:5:38) (running) created at:
  github.com/tikv/pd/pkg/utils/syncutil.TestNegativeDelta()
      /home/runner/work/pd/pd/pkg/utils/syncutil/flexible_wait_group_test.go:81 +0x205
  testing.tRunner()
      /opt/hostedtoolcache/go/1.21.1/x64/src/testing/testing.go:1595 +0x2[38](https://github.com/tikv/pd/actions/runs/6252103050/job/16974605966?pr=7107#step:5:39)
  testing.(*T).Run.func1()
      /opt/hostedtoolcache/go/1.21.1/x64/src/testing/testing.go:1648 +0x[44](https://github.com/tikv/pd/actions/runs/6252103050/job/16974605966?pr=7107#step:5:45)```

File: pkg/utils/syncutil/flexible_wait_group.go [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+}
+
+// Done decrements the FlexibleWaitGroup counter by one.
+func (fwg *FlexibleWaitGroup) Done() {
+	fwg.Add(-1)
+}
+
+// Wait blocks until the FlexibleWaitGroup counter is zero or negative.
+func (fwg *FlexibleWaitGroup) Wait() {
+	fwg.Lock()
```
Comment (nolouch) [inline pkg/utils/syncutil/flexible_wait_group.go:57]: will it always hold the lock?

Comment (lhy1024) [reply pkg/utils/syncutil/flexible_wait_group.go:57]: No, fwg.cond.Wait() will work

**Pull request closed without merging**